File: tests/algorithms/transpiler/FS/matrix/count_negative_numbers_in_sorted_matrix.fs
// Generated 2025-08-17 13:19 +0700

exception Break
exception Continue

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.10g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
let _floordiv (a:int) (b:int) : int =
    let q = a / b
    let r = a % b
    if r <> 0 && ((a < 0) <> (b < 0)) then q - 1 else q
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec generate_large_matrix () =
    let mutable __ret : int array array = Unchecked.defaultof<int array array>
    try
        let mutable result: int array array = Array.empty<int array>
        let mutable i: int = 0
        while i < 1000 do
            let mutable row: int array = Array.empty<int>
            let mutable j: int = 1000 - i
            while j > ((-1000) - i) do
                row <- Array.append row [|j|]
                j <- j - 1
            result <- Array.append result [|row|]
            i <- i + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and find_negative_index (arr: int array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable arr = arr
    try
        let mutable left: int = 0
        let mutable right: int = (Seq.length (arr)) - 1
        if (Seq.length (arr)) = 0 then
            __ret <- 0
            raise Return
        if (_idx arr (int 0)) < 0 then
            __ret <- 0
            raise Return
        while left <= right do
            let mid: int = _floordiv (int (left + right)) (int 2)
            let num: int = _idx arr (int mid)
            if num < 0 then
                if mid = 0 then
                    __ret <- 0
                    raise Return
                if (_idx arr (int (mid - 1))) >= 0 then
                    __ret <- mid
                    raise Return
                right <- mid - 1
            else
                left <- mid + 1
        __ret <- Seq.length (arr)
        raise Return
        __ret
    with
        | Return -> __ret
and count_negatives_binary_search (grid: int array array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable grid = grid
    try
        let mutable total: int = 0
        let mutable bound: int = Seq.length (_idx grid (int 0))
        let mutable i: int = 0
        while i < (Seq.length (grid)) do
            let mutable row: int array = _idx grid (int i)
            let idx: int = find_negative_index (Array.sub row 0 (bound - 0))
            bound <- idx
            total <- total + idx
            i <- i + 1
        __ret <- int (((int64 (Seq.length (grid))) * (int64 (Seq.length (_idx grid (int 0))))) - (int64 total))
        raise Return
        __ret
    with
        | Return -> __ret
and count_negatives_brute_force (grid: int array array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable grid = grid
    try
        let mutable count: int = 0
        let mutable i: int = 0
        while i < (Seq.length (grid)) do
            let mutable row: int array = _idx grid (int i)
            let mutable j: int = 0
            while j < (Seq.length (row)) do
                if (_idx row (int j)) < 0 then
                    count <- count + 1
                j <- j + 1
            i <- i + 1
        __ret <- count
        raise Return
        __ret
    with
        | Return -> __ret
and count_negatives_brute_force_with_break (grid: int array array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable grid = grid
    try
        let mutable total: int = 0
        let mutable i: int = 0
        try
            while i < (Seq.length (grid)) do
                try
                    let mutable row: int array = _idx grid (int i)
                    let mutable j: int = 0
                    try
                        while j < (Seq.length (row)) do
                            try
                                let number: int = _idx row (int j)
                                if number < 0 then
                                    total <- total + ((Seq.length (row)) - j)
                                    raise Break
                                j <- j + 1
                            with
                            | Continue -> ()
                            | Break -> raise Break
                    with
                    | Break -> ()
                    | Continue -> ()
                    i <- i + 1
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret <- total
        raise Return
        __ret
    with
        | Return -> __ret
let grid: int array array = generate_large_matrix()
let test_grids: int array array array = [|[|[|4; 3; 2; -1|]; [|3; 2; 1; -1|]; [|1; 1; -1; -2|]; [|-1; -1; -2; -3|]|]; [|[|3; 2|]; [|1; 0|]|]; [|[|7; 7; 6|]|]; [|[|7; 7; 6|]; [|-1; -2; -3|]|]; grid|]
let mutable results_bin: int array = Array.empty<int>
let mutable i: int = 0
while i < (Seq.length (test_grids)) do
    results_bin <- Array.append results_bin [|(count_negatives_binary_search (_idx test_grids (int i)))|]
    i <- i + 1
ignore (printfn "%s" (_str (results_bin)))
let mutable results_brute: int array = Array.empty<int>
i <- 0
while i < (Seq.length (test_grids)) do
    results_brute <- Array.append results_brute [|(count_negatives_brute_force (_idx test_grids (int i)))|]
    i <- i + 1
ignore (printfn "%s" (_str (results_brute)))
let mutable results_break: int array = Array.empty<int>
i <- 0
while i < (Seq.length (test_grids)) do
    results_break <- Array.append results_break [|(count_negatives_brute_force_with_break (_idx test_grids (int i)))|]
    i <- i + 1
ignore (printfn "%s" (_str (results_break)))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/matrix/count_negative_numbers_in_sorted_matrix.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _intdiv($a, $b) {
    if ($b === 0 || $b === '0') {
        throw new DivisionByZeroError();
    }
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        $q = bcdiv($sa, $sb, 0);
        $rem = bcmod($sa, $sb);
        $neg = ((strpos($sa, '-') === 0) xor (strpos($sb, '-') === 0));
        if ($neg && bccomp($rem, '0') != 0) {
            $q = bcsub($q, '1');
        }
        return intval($q);
    }
    $ai = intval($a);
    $bi = intval($b);
    $q = intdiv($ai, $bi);
    if ((($ai ^ $bi) < 0) && ($ai % $bi != 0)) {
        $q -= 1;
    }
    return $q;
}
function generate_large_matrix() {
  global $grid, $results_bin, $results_break, $results_brute, $test_grids;
  $result = [];
  $i = 0;
  while ($i < 1000) {
  $row = [];
  $j = 1000 - $i;
  while ($j > (-1000 - $i)) {
  $row = _append($row, $j);
  $j = $j - 1;
};
  $result = _append($result, $row);
  $i = $i + 1;
};
  return $result;
}
function find_negative_index($arr) {
  global $grid, $i, $results_bin, $results_break, $results_brute, $test_grids;
  $left = 0;
  $right = count($arr) - 1;
  if (count($arr) == 0) {
  return 0;
}
  if ($arr[0] < 0) {
  return 0;
}
  while ($left <= $right) {
  $mid = _intdiv(($left + $right), 2);
  $num = $arr[$mid];
  if ($num < 0) {
  if ($mid == 0) {
  return 0;
};
  if ($arr[$mid - 1] >= 0) {
  return $mid;
};
  $right = $mid - 1;
} else {
  $left = $mid + 1;
}
};
  return count($arr);
}
function count_negatives_binary_search($grid) {
  global $results_bin, $results_break, $results_brute, $test_grids;
  $total = 0;
  $bound = count($grid[0]);
  $i = 0;
  while ($i < count($grid)) {
  $row = $grid[$i];
  $idx = find_negative_index(array_slice($row, 0, $bound));
  $bound = $idx;
  $total = $total + $idx;
  $i = $i + 1;
};
  return (count($grid) * count($grid[0])) - $total;
}
function count_negatives_brute_force($grid) {
  global $results_bin, $results_break, $results_brute, $test_grids;
  $count = 0;
  $i = 0;
  while ($i < count($grid)) {
  $row = $grid[$i];
  $j = 0;
  while ($j < count($row)) {
  if ($row[$j] < 0) {
  $count = $count + 1;
}
  $j = $j + 1;
};
  $i = $i + 1;
};
  return $count;
}
function count_negatives_brute_force_with_break($grid) {
  global $results_bin, $results_break, $results_brute, $test_grids;
  $total = 0;
  $i = 0;
  while ($i < count($grid)) {
  $row = $grid[$i];
  $j = 0;
  while ($j < count($row)) {
  $number = $row[$j];
  if ($number < 0) {
  $total = $total + (count($row) - $j);
  break;
}
  $j = $j + 1;
};
  $i = $i + 1;
};
  return $total;
}
$grid = generate_large_matrix();
$test_grids = [[[4, 3, 2, -1], [3, 2, 1, -1], [1, 1, -1, -2], [-1, -1, -2, -3]], [[3, 2], [1, 0]], [[7, 7, 6]], [[7, 7, 6], [-1, -2, -3]], $grid];
$results_bin = [];
$i = 0;
while ($i < count($test_grids)) {
  $results_bin = _append($results_bin, count_negatives_binary_search($test_grids[$i]));
  $i = $i + 1;
}
echo rtrim(_str($results_bin)), PHP_EOL;
$results_brute = [];
$i = 0;
while ($i < count($test_grids)) {
  $results_brute = _append($results_brute, count_negatives_brute_force($test_grids[$i]));
  $i = $i + 1;
}
echo rtrim(_str($results_brute)), PHP_EOL;
$results_break = [];
$i = 0;
while ($i < count($test_grids)) {
  $results_break = _append($results_break, count_negatives_brute_force_with_break($test_grids[$i]));
  $i = $i + 1;
}
echo rtrim(_str($results_break)), PHP_EOL;
